server/game/cards/22-BtB/Highgarden.js

```diff
@@ -0,0 +1,140 @@
+const range = require('lodash.range');
+
+class Highgarden extends DrawCard {
+    setupCardAbilities(ability) {
+        this.action({
+            title: 'Give characters +STR',
+            condition: () => this.controller.hand.length > 0,
+            cost: ability.costs.kneelSelf(),
+            handler: () => {
+                this.game.promptForSelect(this.controller, {
+                    activePromptTitle: 'Reveal up to 3 cards',
+                    source: this,
+                    numCards: 3,
+                    mode: 'upTo',
+                    cardCondition: card => card.controller === this.controller && card.location === 'hand' && card.isFaction('tyrell') && card.getType() === 'character',
+                    onSelect: (player, cards) => this.revealSelect(player, cards),
+                    onCancel: (player) => this.revealCancel(player)
+                });
+            }
+        });
+        this.action({
+            title: 'Return character to hand',
+            phase: 'challenge',
+            limit: ability.limit.perPhase(1),
+            target: {
+                cardCondition: card => card.location === 'play area' && card.controller === this.controller && 
+                    card.getType() === 'character' && card.isFaction('tyrell') && card.isUnique()
+            },
+            handler: context => {
+                context.target.owner.returnCardToHand(context.target);
+                this.game.addMessage('{0} uses {1} to return {2} to their hand', this.controller, this, context.target);
+            }
+        });
+    }
+
+    revealSelect(player, revealed) {
+        this.game.addMessage('{0} kneels {1} to reveal {2} from their hand', player, this, revealed);
+
+        this.game.resolveGameAction(GameActions.simultaneously(
+            revealed.map(card => GameActions.revealCard({ card }))
+        ));
+
+        let numRevealed = revealed.length;
+
+        this.game.promptForSelect(this.controller, {
+            activePromptTitle: `Select up to ${numRevealed} characters`,
+            source: this,
+            numCards: numRevealed,
+            mode: 'upTo',
+            cardCondition: card => card.getType() === 'character' && card.location === 'play area',
+            onSelect: (player, cards) => this.charactersSelect(player, cards, numRevealed),
+            onCancel: (player) => this.charactersCancel(player)
+        });
+        return true;
+    }
+
+    revealCancel(player) {
+        this.game.addAlert('danger', '{0} did not reveal any cards for {1}', player, this);
+
+        return true;
+    }
+
+    charactersSelect(player, cards, numRevealed) {
+        this.strCharacterMap = new Map();
+        this.remainingCards = cards;
+        this.remainingStr = numRevealed * 2;
+        
+        this.game.queueSimpleStep(() => this.calculateNextCharacter(player));
+        
+        this.game.queueSimpleStep(() => {
+            for(let [str, cards] of this.strCharacterMap) {
```

Comment:
This definitely works, but it's perhaps conceptually a bit strange to use a map for this. I think a simple array with plain objects (e.g. `{ str: 2, card: Arya }`) would be more straightforward. (and probably closer to how it would work if it were a real prompt class)

Reply:
Fair enough, however I decided to slightly alter this and use the same logic we have for many dynamic effects, where its stored in an object tying the effect to each card uuid affected. In this case, the property name is the STR increase (2 or 4), and the value is an array of characters its affected. This is part of the 63dd48b update; hopefully it suffices.